File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for(int i=0; i<s.length(); i++){
            char character = s.charAt(i);

            if(character == '(' || character == '[' || character == '{'){
                stack.push(character);
            }
            else{
                if(stack.isEmpty()){
                    return false;
                }   

                char top = stack.pop();
                if ((character == ')' && top != '(') ||
                    (character == ']' && top != '[') ||
                    (character == '}' && top != '{')) {
                    return false;
                }
            }
        }
        
        return stack.isEmpty();
    }
}

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for char in s:
            if char == "(" or char == "{" or char == "[":
                stack.append(char)

            elif char == ")" or char == "}" or char == "]":
                if not stack:
                    return False

                if char == ")" and stack[-1] == "(":
                    stack.pop()
                elif char == "}" and stack[-1] == "{":
                    stack.pop()
                elif char == "]" and stack[-1] == "[":
                    stack.pop()
                else:
                    return False
                    
        return not stack

        



       